### PR TITLE
refactor thrift serialization's compilation & add support for rpc with thrift

### PR DESCRIPTION
**Pull request by shengofsun**
1. Fix thrift's compilation, currently the thrift library is compiled to a dynamic library on Linux. More works need to be done on Windows.
2. The cmake option "DSN_SERIALIZATION_TYPE" is used to control the serialization method of rDSN. On Linux "thrift" and "dsn" is supported, and on windows currently only "dsn" is supported. 
3. Add serialization type option in "run.sh". For example, "./run.sh build -s thrift" will build the system with thrift serialization.
4. Modify the message_parser to support the thrift header parser. Thus we can write replication client with other language.

### CODE REVIEW

**Comment by ykwd** [comment]
@shengofsun  After this commit I may merge thrift's compilation and installation code from branch layer3 to branch master.

**Comment by shengofsun** [comment]
@ykwd There are some points you may have interests: 
1. You may want to add an option in window's "run.cmd", so we can specify which serialization method to use when compiling. 
2. Currently, the thrift project is a tar.gz package(for faster download speed) in my repo. I have a tiny modification on the "build/cmake/ThriftMacros.cmake" to remove the version postfix of library(from libthrift.so.0.9.3 to libthrift.so). If you'd like to use another thrift repo, you may need to take care of this.
3. I delete the "DSN_NOT_USE_DEFAULT_SERIALIZATION" macro, and replace it with "DSN_USE_THRIFT_SERIALIZATION" in rDSN project.

File: include/dsn/c/api_layer1.h [hunk 1]
```diff
@@ -536,7 +536,9 @@ typedef union dsn_msg_context_t
         uint64_t is_request : 1;           ///< whether the RPC message is a request or response
         uint64_t is_forwarded : 1;         ///< whether the msg is forwarded or not
         uint64_t is_replication_needed: 1; ///< whether state replication is needed for this request
-        uint64_t unused : 8;               ///< not used yet
+        uint64_t is_response_in_piece: 1;  ///< set when a response is marshalled in different place
```
Comment (imzhenyu) [inline include/dsn/c/api_layer1.h]: @shengofsun any more detailed comments about these flags? the current comments are not easy to follow.

File: include/dsn/cpp/rpc_stream.h [hunk 2]
```diff
@@ -38,6 +38,7 @@
 # include <dsn/cpp/utils.h>
 # include <dsn/service_api_c.h>
 # include <dsn/cpp/auto_codes.h>
+# include <dsn/internal/rpc_message.h>
```
Comment (imzhenyu) [inline include/dsn/cpp/rpc_stream.h:41]: this is forbidden as dev layers cannot access internal data structures in core. the layout of the structure in core is not guaranteed to be the same as the one compiled right now in the app layer.

File: include/dsn/cpp/rpc_stream.h [hunk 3]
```diff
@@ -97,6 +98,9 @@ namespace dsn
         {
             _last_write_next_committed = true;
             _last_write_next_total_size = 0;
+#ifdef DSN_USE_THRIFT_SERIALIZATION
```
Comment (imzhenyu) [inline include/dsn/cpp/rpc_stream.h]: why we need to couple rpc_stream with thrift? we also have other encoder/decoders being developed like protoc and bond ...

File: include/dsn/c/api_layer1.h (hunk 1 above)
Comment (shengofsun) [inline include/dsn/c/api_layer1.h]: Sorry, my bad. I will fix this.

File: include/dsn/idl/thrift_helper.h [hunk 4]
```diff
@@ -284,12 +285,44 @@ namespace dsn {
...
+    template<typename T>
+    inline ::apache::thrift::protocol::TType get_thrift_type(const T&)
+    {
+        return ::apache::thrift::protocol::T_STRUCT;
+    }
+
     template<typename T>
     void marshall(binary_writer& writer, const T& val)
     {
         boost::shared_ptr< ::dsn::binary_writer_transport> transport(new ::dsn::binary_writer_transport(writer));
```
Comment (imzhenyu) [inline include/dsn/idl/thrift_helper.h]: this is my bad. But can we remove new/delete here but use on-stack vars here instead to improve the performance?

File: include/dsn/cpp/rpc_stream.h [hunk 5]
```diff
@@ -97,6 +98,9 @@ namespace dsn
...
+            _p_value_id = &(reinterpret_cast<dsn::message_ex*>(msg)->_value_id);
```
Comment (imzhenyu) [inline include/dsn/cpp/rpc_stream.h]: Are these specially added parameters for message parsing etc.? If yes, can we put them into the message parser instead? As one message parser is used for one  In that case, things will be much simpler.

File: src/dist/replication/lib/replica_stub.cpp [hunk 6]
```diff
@@ -383,6 +383,11 @@ void replica_stub::on_client_write(dsn_message_t request)
     write_request_header hdr;
     ::unmarshall(request, hdr);
 
+    dsn_msg_options_t opt;
+    dsn_msg_get_options(request, &opt);
+    opt.context.u.is_response_in_piece = 1;
```
Comment (imzhenyu) [inline src/dist/replication/lib/replica_stub.cpp]: what does this flag mean?

File: src/dist/replication/meta_server/meta_service.cpp [hunk 7]
```diff
@@ -231,29 +231,33 @@ void meta_service::start_load_balance()
...
 }
 
-bool meta_service::check_primary(dsn_message_t req)
+int meta_service::check_primary(dsn_message_t req)
 {
     if (!_failure_detector->is_primary())
     {
+        dsn_msg_options_t options;
+        dsn_msg_get_options(req, &options);
+        if ( options.context.u.is_forward_disabled )
```
Comment (imzhenyu) [inline src/dist/replication/meta_server/meta_service.cpp:240]: Is this configuration (as well as the above xxx_in_one_piece) for one single message? or for one kind of rpc message (e.g., for all RPC_PREPARE messages)? If the latter case is true, we should remove these flags, and make them available in task_spec  (i.e., configurable in config.ini). For the former case, would you please illustrate some cases so we can do assessment about why this is necessary?

**Comment by imzhenyu** [comment]
One more issue, we use our changed thrift version to do the code generation (added --out rDSN). Therefore the external project should use imzhenyu/thrift or anyone derived from that repo to avoid possible inconsistencies due to different versions of thrifts in our code generators and the externally referenced header/libs.

**Comment by ykwd** [comment]
I have upgraded the project of imzhenyu/thrift to version 0.9.3, which is identical to the version of master's thrift. And I have also omitted the postfix of library(for  both windows and linux), just the same as master branch. And I have also modified the thrift so that it won't generate skeleton.cpp.

**Comment by shengofsun** [comment]
@imzhenyu let me answer this in detail.

Well, I do all these things because we want to write a Java client without the JNI calls to rDSN's runtime.

Several reasons for this: 
1. Currently rDSN have bug and may dump core when using JNI.
2. Proxy way is not a perfect solution.

In xiaomi's case, we want to switch an existing service from HBase to RRDB. During the switching, we'd like to do the read/write simultaneously both to HBase and RRDB. For this case, a Java client is a very natural way.

So I generate some Java code with Thrift and let it communicate with the RRDB cluster directly. This isn't very easy for three reasons: 

(a) Our rDSN's rpc has a custom message_header.
(b) Our serialization protocol isn't fully compatible.
(c) Our rDSN message can be forwarded. So I'd like to disable it when I find this is a message from Thrift client.

Then let me explain (b). In thrift, all message are wrapped in a struct. And each field in the struct is assigned with a ID starting from 1. But in rDSN, we can easily append anything to a message with "marshall". A typical situation is the response for the client read/write:

```
if (the replication layer have error):
   then marshal(error_code)
else:
   marshall(error_ok)
   marshall(replication_app_data) 
```

In order to make this work well with thrift, I need to reconstruct the error_code and optional replication_app_data with a Thrift structure. 

And the response_in_piece is work for this. 

In addition, in order to calculate the IDs for these fields, I have to store a integer in the class "message_ex", which increases each time the marshalling is called.

Maybe this is somewhat specific to the Thrift, but I think this isn't too specific. After all, both thrift and protobuf have IDs for the structure's fields, which is a very useful feature for protocol upgrade.

A better way may be changing our marshall interface to "marshall(rpc_stream, field, field_id)". But too much code is influenced.

The last is the message_header thing. I have no choice but to add a message_header on the Thrift client part. This isn't very hard as it's easy to inherit the Thrift's TProtocol and TTransport without modification the Thrift library.

Now the java client is [runnable.](https://github.com/shengofsun/jardb.git). Currently only works for rrdb. But it isn't hard to write a code generator to support all kinds of IDL. And I think this is a better way to support multiple language if we only support the client for different language.

**Comment by imzhenyu** [comment]
Thanks @shengofsun . I'm afraid as far as I understand about those thrift/protoc etc., you may have a misunderstanding here. For those numbers 1,2,3,4... before the field or argument types, they are not 1-based indices, but more like an index name, i.e., 1 always stands for this particular field, which can be removed in future versions of the data structure while 1 is preserved and cannot be used for others anymore, in order to make sure cross-version parsing compatibility. That's said, the marshalled data is a self-explained data structure where 1,2,3,4 define the schemas. Therefore, you cannot rely on auto-increment for those value, but instead consider them part of the data structure themselves. For the issue you raised, we can fix the response type in replication, i.e., make sure they are always consistent in terms of their type. This is much easier once our layer2 application model is made consistent with layer 1, i.e., both server and client does not change for layer 1 or layer 2. 

As for the message header, I think in the very old version of our rDSN, we already have the thrift-message-parser that implements inter-communication between thrift and rDSN services. The main idea is to generate a message header from the other side on demand.

**Comment by shengofsun** [comment]
Well, I know 1,2,3,4... are index names but not 1-based indices. And they are assigned in the IDL file. So whether it is auto increment depends on how we assigned it in the IDL files.  

What I mean is like this(for the simple_kv case): 

if we have an original IDL file like this:

```
service simple_kv
{
    string read(1:string key);
}
```

we can write a tool to process the IDL file to another file

```
struct replication_read_response
{
    1:dsn.error_code ec;
    2:optional string application_result;
}

service simple_kv
{
    replication_read_response read(1: write_request_header, 2: key);
}
```

as the replication_read_response is generated by our preprocessor, so we can make it auto increment. 

This is a little hack, but we can easily write clients for different languages and keep the server C++.

**Comment by shengofsun** [comment]
Of course we have to write the query meta, query partition server things for every language. But only once.

**Comment by imzhenyu** [comment]
This is somehow quite hacky, and there are a number of issues.
- wrong coupling of the internal data structure of the specific encoding/decoding methodology and the general message_ex etc. 

Our message ex should only care about msg-header + msg-body, where msg-body is a binary data which is defined by the upper encoder/decoder. In current case, thrift-specific implementations are put into the internal of message-ex, which is not appropriate and be difficult to be used by other IDL generators (see next). 
- the 1/2/3 naming of the service call parameters is unique to thrift, which does not exist in other IDL generators such as in protoc. 

What is really there is something like this:

```
struct service_call_foo_args
{
  1: param1-type param1-name;
  2: param2-type param2-name;
};
```

This is what protoc requires explicitly in .proto file, and thrift introduces a syntactic sugar by automatically generating this. As a result, we don't really expect any 1/2/3 parsing during message-ex parsing for protoc/thrift but that should be all handled by the `unmarsall`` function derived from what protoc/thrift generates. 
- the problem is about (b), as you mentioned above, where we don't really have a consistent response type for certain rpc calls. 

This is indeed a bad practice, and one way (as we did before in Bing) is to adopt blob for the application level response. Therefore the rpc response of replicated service calls are:

```
struct replicated_app_response
{
  1: error-code err;
  2: blob app-response;
}
```

Developers further use other unmarshall functions to get the typed app response from app-response.

This problem will be gone after we merge the same-client-for-layer1-and-layer2 into master branch., as you may find here: 
https://github.com/imzhenyu/rDSN/commit/d28e41ae8d92377b19b4826a6407f75bf4c37b3f

You may try it on branch layer2 or layer3. 

I'll discuss the issues about JNI and native java client next, as this msg is too long and my browser getting response-slow:)

**Comment by imzhenyu** [comment]
Given we have same-client-for-layer1-and-layer2, a JNI client lib atop of dsn.core.so will be possible. In fact, we have done this before for the layer 1 though right now it is broken, see repo  https://github.com/rDSN-Projects/rDSN.Java .

I'm interested in the coredump of your JNI implementation though - maybe we should find a chance to sync about the possible issues there.

Another way is to write native Java client lib for rrdb, as you have already did. The problem is as you mentioned (a). how to handle the inconsistent message header, as the message body part is already consistent using thrift encoding/decoding. In our old practice, we send a common thrift request msg to rdsn, and rdsn uses a thrift-message-parse to automatically synthesize a message-ex::message-header from thrift's message header. this is possible when only the rpc-code is important while all the other fields can be set to be default. Apparently, this is not the case, as we may also have other important fields such as hash. Right now, I don't think I have a solution for this. 

For issue (c), I think it is necessary if you use native Java client lib. 

In all, I would suggest putting your changes into a dedicated branch as being a temporal solution, as the solution is hacky and non-stable in the future. I think we should prefer the JNI approach, which avoids doing those hacky things and allows rDSN to remain orthogonal to all IDL methods (thrift/protoc/bond). 

What do you think?

**Pull request closed without merging**